### PR TITLE
Implement exchange support for linked-definition sections

### PR DESCRIPTION
**Pull request by paulhuggett**
This changes the pstore API to record the compilation digest and index of the definition referenced by an entry in a linked-definition section. Previously, we simply recorded a pointer to the definition: that’s perfect for most consumers but not sufficient to enable the pstore export code to represent the target object. The structure now contains the referenced compilation digest and the index of the relevant definition in addition to the original pointer.

- Implemented export and import of linked-definition sections.
- Eliminated the import_ and export_ prefixes in favour of using C++ namespaces (since that’s what they’re for)!
- Introduced an import “context” which carries the parser stack, a pointer to the target database, and (introduced for this work) a collection of “patch” objects which turn the imported digest/index pair into the final address. This patching must be performed _after_ the fragments have been imported because the referenced compilation may not have been imported when the reference is consumed.

Whilst working on the compiler part of this change, I spotted a probable bug caused by calling code failing to keep the pointer returned from pstore alive. That bug was confirmed using the PSTORE_ALWAYS_SPANNING setting. This provoked a change to the `get_sstring_view()` helper function to hopefully reduce the probability of a similar bug in future.

### CODE REVIEW

File: include/pstore/exchange/import_compilation.hpp [hunk 1]
```diff
@@ -62,254 +62,247 @@
...
+            //* / _` / -_)  _| | ' \| |  _| / _ \ ' \  / _ \ '_ \| / -_) _|  _| *
+            //* \__,_\___|_| |_|_||_|_|\__|_\___/_||_| \___/_.__// \___\__|\__| *
+            //*                                                |__/             *
+            //-MARK: definition object
+            class definition_object final : public rule {
+            public:
+                definition_object (not_null<context *> ctxt,
+                                   not_null<definition::container *> definitions,
+                                   not_null<name_mapping const *> names,
+                                   fragment_index_pointer const & fragments);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_compilation.hpp:124]: Missing `const` for parameters?

definition_object (not_null<context *> **const** ctxt,
                                   not_null<definition::container *> **const** definitions,
                                   not_null<name_mapping const *> **const** names,
                                   fragment_index_pointer const & fragments);

File: include/pstore/exchange/import_debug_line_header.hpp [hunk 2]
```diff
@@ -59,95 +59,98 @@
...
-            if (!from_base64 (std::begin (s), std::end (s), std::back_inserter (data))) {
-                return import_error::bad_base64_data;
+        namespace import {
+
+            template <typename TransactionLock>
+            class debug_line_index final : public rule {
+            public:
+                using transaction_pointer = gsl::not_null<transaction<TransactionLock> *>;
+
+                debug_line_index (not_null<context *> ctxt, transaction_pointer transaction);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_debug_line_header.hpp:69]: Missing `const`?

debug_line_index (not_null<context *> **const** ctxt, transaction_pointer **const** transaction);

File: include/pstore/exchange/import_fixups.hpp [hunk 3]
```diff
@@ -54,153 +54,154 @@
...
-                section_name (parse_stack_pointer const stack,
-                              not_null<pstore::repo::section_kind *> const section)
-                        : import_rule (stack)
-                        , section_{section} {}
-                section_name (section_name const &) = delete;
-                section_name (section_name &&) noexcept = delete;
+                using fixups_pointer = not_null<std::vector<repo::external_fixup> *>;
+
+                external_fixup (not_null<context *> ctxt, not_null<name_mapping const *> names,
+                                fixups_pointer fixups);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_fixups.hpp:139]: Missing `const`?

external_fixup (not_null<context *> **const** ctxt, not_null<name_mapping const *> **const** names,
                         fixups_pointer **const** fixups);

File: include/pstore/exchange/import_fixups.hpp [hunk 4]
```diff
@@ -54,153 +54,154 @@
...
+            //* | | ' \  _/ -_) '_| ' \/ _` | | |  _| \ \ / || | '_ \ *
+            //* |_|_||_\__\___|_| |_||_\__,_|_| |_| |_/_\_\\_,_| .__/ *
+            //*                                                |_|    *
+            //-MARK:internal fixup
+            class internal_fixup final : public rule {
+            public:
+                using fixups_pointer = not_null<std::vector<repo::internal_fixup> *>;
+
+                internal_fixup (not_null<context *> ctxt, not_null<name_mapping const *> names,
+                                fixups_pointer fixups);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_fixups.hpp:102]: Missing `const`?

internal_fixup (not_null<context *> **const** ctxt, not_null<name_mapping const *> **const** names,
                        fixups_pointer **const** fixups);

File: include/pstore/exchange/import_fragment.hpp [hunk 5]
```diff
@@ -61,167 +61,229 @@
...
+            //*              |___/                                          *
+            //-MARK: fragment index
+            template <typename TransactionLock>
+            class fragment_index final : public rule {
+            public:
+                using transaction_pointer = not_null<transaction<TransactionLock> *>;
+                using names_pointer = not_null<name_mapping const *>;
+
+                fragment_index (not_null<context *> ctxt, transaction_pointer transaction,
+                                names_pointer names);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_fragment.hpp:225]: Missing `const`?

fragment_index (not_null<context *> **const** ctxt, transaction_pointer **const** transaction,
                           names_pointer **const** names);

File: include/pstore/exchange/import_names_array.hpp [hunk 6]
```diff
@@ -56,67 +56,67 @@
...
-        class names_array_members final : public import_rule {
-        public:
-            using transaction_pointer = not_null<transaction<TransactionLock> *>;
-            using names_pointer = not_null<import_name_mapping *>;
+            template <typename TransactionLock>
+            class names_array_members final : public rule {
+            public:
+                names_array_members (not_null<context *> ctxt,
+                                     not_null<transaction<TransactionLock> *> transaction,
+                                     not_null<name_mapping *> names);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_names_array.hpp:66]: Missing `const`?

names_array_members (not_null<context *> **const** ctxt,
                                     not_null<transaction<TransactionLock> *> **const** transaction,
                                     not_null<name_mapping *> **const** names);

File: include/pstore/exchange/import_transaction.hpp [hunk 7]
```diff
@@ -62,159 +62,160 @@
...
+
+            //*  _                             _   _                                    *
+            //* | |_ _ _ __ _ _ _  ___ __ _ __| |_(_)___ _ _    __ _ _ _ _ _ __ _ _  _  *
+            //* |  _| '_/ _` | ' \(_-</ _` / _|  _| / _ \ ' \  / _` | '_| '_/ _` | || | *
+            //*  \__|_| \__,_|_||_/__/\__,_\__|\__|_\___/_||_| \__,_|_| |_| \__,_|\_, | *
+            //*                                                                   |__/  *
+            template <typename TransactionLock>
+            class transaction_array final : public rule {
+            public:
+                transaction_array (not_null<context *> ctxt, not_null<name_mapping *> names);
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_transaction.hpp:179]: Missing `const`?

transaction_array (not_null<context *> **const** ctxt, not_null<name_mapping *> **const** names);

File: include/pstore/exchange/import_uuid.hpp [hunk 8]
```diff
@@ -49,26 +49,28 @@
...
+        namespace import {
 
-        class import_uuid_rule final : public import_rule {
-        public:
-            import_uuid_rule (parse_stack_pointer const stack, not_null<uuid *> const v) noexcept;
-            import_uuid_rule (import_uuid_rule const &) = delete;
-            import_uuid_rule (import_uuid_rule &&) = delete;
+            class uuid_rule final : public rule {
+            public:
+                uuid_rule (not_null<context *> ctxt, not_null<uuid *> v) noexcept;
```
Comment (MaggieYingYi) [inline include/pstore/exchange/import_uuid.hpp:56]: Missing `const`?

uuid_rule (not_null<context *> **const** ctxt, not_null<uuid *>  **const** v) noexcept;

File: include/pstore/mcrepo/linked_definitions_section.hpp [hunk 9]
```diff
@@ -79,12 +90,37 @@ namespace pstore {
         /// record.
         class linked_definitions : public section_base {
         public:
-            using iterator = typed_address<compilation_member> *;
-            using const_iterator = typed_address<compilation_member> const *;
-
-            template <typename Iterator, typename = typename std::enable_if<std::is_same<
-                                             typename std::iterator_traits<Iterator>::value_type,
-                                             typed_address<compilation_member>>::value>::type>
+            struct value_type {
```
Comment (MaggieYingYi) [inline include/pstore/mcrepo/linked_definitions_section.hpp:93]: Could you add some documents for the struct of value_type?

**Review by MaggieYingYi — COMMENTED** [review]
Can export.hpp and export_emit.hpp be combined to a single header?

File: include/pstore/exchange/export.hpp [hunk 10]
```diff
@@ -49,10 +49,12 @@
 
 namespace pstore {
     namespace exchange {
+        namespace export_ns {
```
[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/export.hpp:55]: `export` is a keyword in C++ meaning that we’re not allowed to use it as the name of a namespace. In this special case, I’ve added the `_ns` suffix to work around this limitation. I don’t want to add it in other cases where the restriction is not present.

I added a comment to explain this in commit 21484b08c080f7103acea3d2af8d3bcf3885c00b.

**Comment by paulhuggett** [comment]
> Can export.hpp and export_emit.hpp be combined to a single header?

Not really. export.hpp is the top-level entry point for export whereas export_emit.hpp has the low-level functions for writing strings, arrays, and so on.

File: include/pstore/exchange/export.hpp [hunk 11]
```diff
@@ -49,10 +49,12 @@
...
 
-        void export_database (database & db, export_ostream & os, bool comments);
+            void emit_database (database & db, ostream & os, bool comments);
```
[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/export.hpp:57]: > Could we remove emit_ from functions which are starting with emit_ in the export_ns namespace?

We could, and in an earlier version of the code that’s exactly what I did. 

However, having the names of these functions containing no verb was very peculiar indeed. For instance, the intent of `emit_database()` is clear: it emits a database. OTOH, it’s not at all clear what a call to `database()` will actually do.

File: include/pstore/exchange/import_compilation.hpp [hunk 12]
```diff
@@ -62,254 +62,247 @@
...
+            //* \__,_\___|_| |_|_||_|_|\__|_\___/_||_| *
+            //*                                        *
+            //-MARK: definition
+            class definition final : public rule {
+            public:
+                using container = std::vector<repo::compilation_member>;
+
+                definition (not_null<context *> ctxt, not_null<container *> definitions,
+                            not_null<name_mapping const *> names,
+                            fragment_index_pointer const & fragments);
```
[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/import_compilation.hpp:81]: No. This is a _declaration_ of the `exchange::definition` ctor. Adding const to formal parameters of a declaration (as opposed to a definition) will draw a warning from (IIRC) clang-tidy about the const having no effect. You’ll notice that the const _is_ present in the definition of `exchange::definition` in import_compilation.cpp.

File: include/pstore/exchange/import_compilation.hpp [hunk 13]
```diff
@@ -62,254 +62,247 @@
...
+            //* / _/ _ \ '  \| '_ \ | / _` |  _| / _ \ ' \(_-< | | ' \/ _` / -_) \ / *
+            //* \__\___/_|_|_| .__/_|_\__,_|\__|_\___/_||_/__/ |_|_||_\__,_\___/_\_\ *
+            //*              |_|                                                     *
+            //-MARK: compilation index
+            template <typename TransactionLock>
+            class compilations_index final : public rule {
+            public:
+                compilations_index (not_null<context *> ctxt,
+                                    gsl::not_null<transaction<TransactionLock> *> transaction,
+                                    gsl::not_null<name_mapping const *> names);
```
[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/import_compilation.hpp:248]: No (as above).

File: include/pstore/exchange/import_fragment.hpp (hunk 5 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_fragment.hpp:225]: No (as above).

File: include/pstore/exchange/import_names_array.hpp (hunk 6 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_names_array.hpp:66]: No (as above).

File: include/pstore/exchange/import_transaction.hpp (hunk 7 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_transaction.hpp:179]: No (as above).

File: include/pstore/exchange/import_uuid.hpp (hunk 8 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_uuid.hpp:56]: No (as above).

File: include/pstore/exchange/export_names.hpp [hunk 14]
```diff
@@ -55,45 +55,52 @@
...
 
-        class export_name_mapping {
-        public:
-            void add (address addr);
+            //*                                             _            *
+            //*  _ _  __ _ _ __  ___   _ __  __ _ _ __ _ __(_)_ _  __ _  *
+            //* | ' \/ _` | '  \/ -_) | '  \/ _` | '_ \ '_ \ | ' \/ _` | *
+            //* |_||_\__,_|_|_|_\___| |_|_|_\__,_| .__/ .__/_|_||_\__, | *
+            //*                                  |_|  |_|         |___/  *
+            class name_mapping {
```
[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/export_names.hpp:65]: Done in commit 0dd1b4e3d4d7e6d64d124aef28faef9e677a4c73.

[the remark this answers is not in the document]
Comment (paulhuggett) [reply include/pstore/exchange/export_names.hpp:69]: > Done in commit [0dd1b4e](https://github.com/SNSystems/pstore/commit/0dd1b4e3d4d7e6d64d124aef28faef9e677a4c73).

Wrong one. it was commit f77197afa5c40406521db8a873d980f804fed69f.

File: include/pstore/mcrepo/linked_definitions_section.hpp (hunk 9 above)
Comment (paulhuggett) [reply include/pstore/mcrepo/linked_definitions_section.hpp:93]: Done in commit 0dd1b4e3d4d7e6d64d124aef28faef9e677a4c73.

File: include/pstore/exchange/import_compilation.hpp (hunk 1 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_compilation.hpp:124]: No (as above).

File: include/pstore/exchange/import_debug_line_header.hpp (hunk 2 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_debug_line_header.hpp:69]: No (as above).

File: include/pstore/exchange/import_fixups.hpp (hunk 3 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_fixups.hpp:139]: No (as above).

File: include/pstore/exchange/import_fixups.hpp (hunk 4 above)
Comment (paulhuggett) [reply include/pstore/exchange/import_fixups.hpp:102]: No (as above).

**Comment by paulhuggett** [comment]
I think that I’ve responded to all of the queries. Please let me know if I’ve missed anything.

**Review by MaggieYingYi — APPROVED** [review]
Many thanks for all detailed answers.

LGTM!

**Comment by paulhuggett** [comment]
> > Can export.hpp and export_emit.hpp be combined to a single header?
> 
> Not really. export.hpp is the top-level entry point for export whereas export_emit.hpp has the low-level functions for writing strings, arrays, and so on.

Commit 7b020aeb0ada652b1e768a58c5e183e9d6d827e0 adds file-level comments to explain this to future readers.

**Pull request merged**